doc/adr/0007-integrate-performance-testing-into-pipeline.md

```diff
@@ -0,0 +1,147 @@
+# 7. Integrate performance testing into the continuous integration pipeline
+
+Date: 2022-08-31
+
+
+## Status
+
+Proposed
+
+
+## Context
+
+We released memory-intensive code to production on July 2022, which caused a long outage.
+
+In the early days, before June 2021, the service discarded performance testing based on the anticipated
+number of referrals per person was not expected to be above a couple of dozen referrals.
+
+Later, performance problems arose from unbounded pagination. In June 2022, we rolled out pagination to mitigate
+and introduced monitoring for slow responses.
+
+The above led to a gap: we could detect slowing down over time, but **we cannot detect slowing down caused by
+changes to the service**.
+
+
+## Decision
+
+1. Introduce read-only<sup>6</sup> performance testing into the continuous integration pipeline:
+
+   - Automatically run before **all** production deployments.
+   - **Block** production deployments if performance tests fail.
+   - The team agrees to **focus on maintaining** the performance tests so they do not fail outside valid reasons.
+      - Flaky tests, environmental issues, data drift, etc. will be prioritised immediately.
+      (Otherwise, the test suite becomes a burden, not help.)
+   - To let critical fixes reach production, we will also provide an **emergency bypass** to allow production
+   deployments to proceed even if the performance test fails.
+      - Can only be used for user-impacting critical fixes. In all other cases, the flaky tests must be fixed first.
+
+2. Extend checklists:
+
+   - Add a refinement checklist about "does this need a performance test".
+   - Add an incident checklist about "would a performance test help detect this problem".
+
+
+### When will we know this succeeded?
+
+- When we revert the fix to the performance problems and see consistent failures on the performance tests.
+- When we re-apply the fix, we see passing tests.
+
+### What does this not solve?
+
+Will not guard against:
+
+- Running out of capacity on concurrent requests.
+- Deadlocks and query locking performance issues, caused by migrations or otherwise.
+- Effects of anomalous peaks of usage.
+- Dependency failures, timeouts, server errors.
+
+
+## Consequences
+
+### 1. Methodology
+
+As of today, we get about around 23-25 requests per second, assets and operational endpoints excluded.
+
+The (Gatling tooling) can define that we want N number of journeys to complete per a specific period.
```

Comment:
I think Gatling just uses journeys per second

Reply:
Fixed in c019f4482514f3dfd40f0bfe3dd96679bc4215e1